Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/J2ObjCParser.kt

```diff
@@ -0,0 +1,218 @@
+package org.jetbrains.kotlin.native.interop.indexer
+
+import org.jetbrains.org.objectweb.asm.ClassReader
+import org.jetbrains.org.objectweb.asm.ClassVisitor
+import org.jetbrains.org.objectweb.asm.MethodVisitor
+import org.jetbrains.org.objectweb.asm.Opcodes
+import org.jetbrains.org.objectweb.asm.Type.getArgumentTypes
+import org.jetbrains.org.objectweb.asm.Type.getReturnType
+import java.util.jar.JarFile
+
+/**
+ * Visits a Java class and builds an ObjCClass
+ */
+class J2ObjCParser: ClassVisitor(Opcodes.ASM5) {
+
+  var className:String = ""
+  val methodDescriptors = mutableListOf<Triple<String, String, Int>>()
+  val parameterNames = mutableListOf<List<String>>()
+
+  override fun visit(version: Int,
+                     access: Int,
+                     name: String,
+                     signature: String?,
+                     superName: String?,
+                     interfaces: Array<out String>?) {
+    className = name
+    super.visit(version, access, name, signature, superName, interfaces)
+  }
+
+  override fun visitMethod(access: Int,
+                           name: String,
+                           descriptor: String,
+                           signature: String?,
+                           exceptions: Array<out String>?): MethodVisitor? {
+
+    val methodBuilder = MethodBuilder(parameterNames)
+    methodDescriptors.add(Triple(name, descriptor, access))
+    return methodBuilder
+  }
+
+  /**
+   * Generates an ObjCClass out of data collected while visiting
+   *
+   * @return An ObjCClass that matches a Java class
+   */
+  fun buildClass(): ObjCClass {
+    val methods = (methodDescriptors zip parameterNames).map { buildClassMethod(it.first.first, it.first.second, it.first.third, it.second)}
+
+    val generatedClass = ObjCClassImpl(
+      name = className,
+      isForwardDeclaration = false,
+      binaryName = null,
+      location = Location(HeaderId("")) // Leaving headerId empty for now
+    )
+    generatedClass.methods.addAll(methods)
+    generatedClass.baseClass = ObjCClassImpl(
+      name = "NSObject",
+      binaryName = null,
+      isForwardDeclaration = false,
+      location = Location(headerId = HeaderId("usr/include/objc/NSObject.h")) // TODO: When implementing inheritance check for proper base class
+    )
+    return generatedClass
+  }
+
+  /**
+   * Creates a ObjCMethod out of method data and parameter names
+   *
+   * @param methodName Name of method
+   * @param methodDesc Descriptor of method
+   * @param access Integer that describes access level of this method
+   * @param paramNames List of parameter names of this method taken from MethodBuilder
+   */
+  private fun buildClassMethod(methodName: String, methodDesc: String, access: Int, paramNames: List<String>): ObjCMethod {
+    if (methodName == "<init>") {
```

Comment:
this could be a property on your data class:
val isConstructor get() = name == "<init>"

Reply:
Done.